contract_compassion/model/contract_group.py

```diff
@@ -0,0 +1,256 @@
+        or BVR """
+        if not context:
+            ctx = dict()
+        else:
+            ctx = context.copy()
+            ctx['lang'] = 'en_US'
+        res = {'value': dict()}
+        payment_term_obj = self.pool.get('account.payment.term')
+        need_bvr_ref_term_ids = payment_term_obj.search(cr, uid, [
+            '|', ('name', 'in', ('Permanent Order', 'BVR')),
+            ('name', 'like', 'multi-months')], context=ctx)
+        lsv_term_ids = payment_term_obj.search(
+            cr, uid, [('name', 'like', 'LSV')], context=ctx)
+        if payment_term_id in need_bvr_ref_term_ids:
+            is_lsv = payment_term_id in lsv_term_ids
+            partner = self.pool.get('res.partner').browse(cr, uid, partner_id,
+                                                          context=context)
+            if partner.ref and (not bvr_ref or is_lsv):
+                res['value'].update({
+                    'bvr_reference': self.compute_partner_bvr_ref(
+                        cr, uid, ids, partner, is_lsv, context)})
+
+        return res
+
+    def on_change_bvr_ref(self, cr, uid, ids, bvr_reference,
+                          context=None):
+        """ Test the validity of a reference number. """
+        is_valid = bvr_reference and bvr_reference.isdigit()
+        if is_valid and len(bvr_reference) == 26:
+            bvr_reference = mod10r(bvr_reference)
+        elif is_valid and len(bvr_reference) == 27:
+            valid_ref = mod10r(bvr_reference[:-1])
+            is_valid = (valid_ref == bvr_reference)
+        else:
+            is_valid = False
+
+        res = dict()
+        if is_valid:
+            res['value'] = {'bvr_reference': bvr_reference}
+        elif bvr_reference:
+            res['warning'] = {'title': _('Warning'),
+                              'message': _('The reference of the partner '
+                                           'has not been set, or is in '
+                                           'wrong format. Please make sure'
+                                           ' to enter a valid BVR '
+                                           'reference for the contract.')}
+        return res
+
+    def compute_partner_bvr_ref(self, cr, uid, ids, partner=None,
+                                is_lsv=False, context=None):
+        """ Generates a new BVR Reference.
+        See file \\nas\it\devel\Code_ref_BVR.xls for more information."""
+        partner = partner or self.browse(cr, uid, ids[0], context).partner_id
+        result = '0' * (9 + (7 - len(partner.ref))) + partner.ref
+        count_groups = str(self.search(
+            cr, uid, [('partner_id', '=', partner.id)], context=context,
+            count=True))
+        result += '0' * (5 - len(count_groups)) + count_groups
+        # Type '0' = Sponsorship
+        result += '0'
+        result += '0' * 4
+
+        if is_lsv:
+            result = '004874969' + result[9:]

```

Comment:
This shouldn't be barcoded. 004874969 is the lsv prefix for compassion account? The bank account should hold this, and if nothing is available, it's time to add it.